File: sdk/java/src/test/java/com/anyvali/SensitiveDataTest.java
package com.anyvali;

import org.junit.jupiter.api.Test;

import java.util.List;
import java.util.Map;

import static com.anyvali.AnyVali.*;
import static org.junit.jupiter.api.Assertions.*;

class SensitiveDataTest {
    @Test
    void encryptedRoundTripAndValidation() {
        var sensitive = new Schema.DescribeOptions().sensitive(true);
        Schema<?> schema = object_(Map.of(
                "name", string(),
                "secret", string().describe("secret", sensitive),
                "profile", object_(Map.of("token", string())).describe("profile", sensitive),
                "aliases", array(string().describe("alias", sensitive))
        ));
        var plain = Map.of(
                "name", "Ada", "secret", "abc",
                "profile", Map.of("token", "xyz"), "aliases", List.of("one", "two")
        );

        Object encrypted = encrypt(schema, plain, (path, value) -> "encrypted:" + value);
        assertTrue(safeParseEncrypted(schema, encrypted).success());
        assertThrows(ValidationError.class,
                () -> encrypt(schema, plain, (path, value) -> "broken"));
        assertNotNull(decrypt(schema, encrypted,
                (path, value) -> switch (String.join(".", path.stream().map(Object::toString).toList())) {
                    case "profile" -> Map.of("token", "xyz");
                    case "aliases.0" -> "one";
                    case "aliases.1" -> "two";
                    default -> "abc";
                }));
    }
}

File: sdk/kotlin/src/test/kotlin/com/anyvali/SensitiveDataTest.kt
package com.anyvali

import org.junit.jupiter.api.Test
import org.junit.jupiter.api.assertThrows
import kotlin.test.assertEquals
import kotlin.test.assertIs

class SensitiveDataTest {
    @Test
    fun `encrypted round trip and validation`() {
        val sensitive = DescribeOptions(sensitive = true)
        val schema = obj(
            mapOf(
                "name" to string(),
                "secret" to string().describe("secret", sensitive),
                "profile" to obj(mapOf("token" to string()), setOf("token")).describe("profile", sensitive),
                "aliases" to array(string().describe("alias", sensitive)),
            ),
            setOf("name", "secret", "profile", "aliases")
        )
        val plain = mapOf(
            "name" to "Ada", "secret" to "abc",
            "profile" to mapOf("token" to "xyz"), "aliases" to listOf("one", "two")
        )
        val encrypted = encrypt(schema, plain) { path, value -> "encrypted:${path.joinToString(".")}:$value" }

        assertIs<ParseResult.Success<*>>(safeParseEncrypted(schema, encrypted))
        assertThrows<ValidationError> { encrypt(schema, plain) { _, _ -> "broken" } }
        assertEquals(plain, decrypt(schema, encrypted) { path, _ ->
            when (path.joinToString(".")) {
                "profile" -> mapOf("token" to "xyz")
                "aliases.0" -> "one"
                "aliases.1" -> "two"
                else -> "abc"
            }
        })
    }
}
